### ISSUE TITLE
Issue with Convolution Backward op

### ISSUE THREAD

**Issue by vivekkhandelwal1** [pytorch/pytorch#97524]
### 🐛 Describe the bug

On running the following python code:
```
grad_out = torch.randn((1, 4, 64, 64))
input_vec = torch.randn((1, 320, 64, 64))
weight = torch.randn((4, 320, 3, 3))

torch.ops.aten.convolution_backward(
            grad_out,
            input_vec,
            weight,
            bias_sizes=[4],
            stride=[1, 1],
            padding=[1, 1],
            dilation=[1, 1],
            transposed=False,
            output_padding=[0, 0],
            groups=1,
            output_mask=[True, True, False])
```
it would give you a tuple with the first two values tensor, and the third value None. On the other hand, if you run it in Torch-MLIR, it generates [IR](https://gist.github.com/ad4b86f64b394c7f1a021d41a8184b25) with the op returning three tensors. It seems there's some bug in the PyTorch run of this op and the torchscript graph generated.

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230323+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.10.9 (main, Dec  7 2022, 01:12:00) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: 11.6.55
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 470.161.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          32
On-line CPU(s) list:             0-31
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      25
Model:                           33
Model name:                      AMD Ryzen 9 5950X 16-Core Processor
Stepping:                        0
Frequency boost:                 enabled
CPU MHz:                         2504.233
CPU max MHz:                     5083.3979
CPU min MHz:                     2200.0000
BogoMIPS:                        6787.07
Virtualization:                  AMD-V
L1d cache:                       512 KiB
L1i cache:                       512 KiB
L2 cache:                        8 MiB
L3 cache:                        64 MiB
NUMA node0 CPU(s):               0-31
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.24.1
[pip3] torch==2.1.0.dev20230323+cpu
[pip3] torchvision==0.16.0.dev20230323+cpu
[conda] Could not collect
```

**Comment by ramiro050** [comment]
To clarify, the issue is that the schema for `convolution_backward` in PyTorch

https://github.com/pytorch/pytorch/blob/35c9ea89fa5b4296218064070c8a380471bebd68/aten/src/ATen/native/native_functions.yaml#L1581

does not match the actual behavior of the op. The schema says the output should be all tensors, but the op will sometimes return `None`s.

**Comment by ramiro050** [comment]
This leads to inconsistent behavior when using regular PyTorch compared to using `torch.jit.script`:

```python
import torch

def forward():
    grad_out = torch.randn((1, 4, 64, 64))
    input_vec = torch.randn((1, 320, 64, 64))
    weight = torch.randn((4, 320, 3, 3))
    result0, _, _ = torch.ops.aten.convolution_backward(
        grad_out, input_vec, weight, bias_sizes=[4], stride=[1, 1],
        padding=[1, 1], dilation=[1, 1], transposed=False,
        output_padding=[0, 0], groups=1, output_mask=[False, True, True])
    return result0 is None

script_forward = torch.jit.script(forward)
print(script_forward.graph)
assert forward() == script_forward()
```

In the above script, `torch.jit.script` simplifies the function to the value `False` because using the schema for `convolution_backward`, `result0` should never be `None`. This results in a disagreement with the actual behavior of the `convolution_backward` op, as shown in the output:

```
graph():
  %46 : bool = prim::Constant[value=0]()
  return (%46)

forward output: True
script_forward output: False
```

**Comment by ramiro050** [comment]
@ngimel, do you know who could help us with this issue?

**Comment by ngimel** [comment]
cc @davidberard98. TorchScript is in maintenance mode and changes to it are generally not made. The difference is indeed that TS can't handle `None` values where schema asks for Tensor but, since TS wasn't used for scripting backwards operations (instead it produced gradients for some subgraphs) directly that issue was never identified and fixed, and likely won't be now. 
Eager mode indeed can sometimes produce `None` where schema calls for `Tensor`

**Comment by vivekkhandelwal1** [comment]
I'm closing this issue since we have added a temporary workaround in Torch-MLIR(https://github.com/llvm/torch-mlir/pull/1773) to get us unblocked for now. We hope that the issue will be addressed in the future.
